### PR TITLE
bus-util: make sure the STOPPING message has been dispatched before e…

### PR DESCRIPTION
**Pull request by lclgo**
…xits

fix: https://github.com/systemd/systemd/issues/29983

### CODE REVIEW

File: src/shared/bus-util.c [hunk 1]
```diff
@@ -130,6 +130,7 @@ int bus_event_loop_with_idle(
                         /* Inform the service manager that we are going down, so that it will queue all
                          * further start requests, instead of assuming we are already running. */
                         sd_notify(false, "STOPPING=1");
+                        sd_notify_barrier(0, 5 * USEC_PER_SEC);
```
Comment (yuwata) [inline src/shared/bus-util.c]: - Please use boolean with short comment: `0` -> `/* unset_environment = */ false`
- Maybe, better to log at least when `-ETIMEDOUT` ? See notify/notify.c.

**Comment by YHNdnzj** [comment]
Why is this needed? From `sd_notify(3)`:

> Note that sd_notify() notifications may be attributed to units correctly only if either the sending process is still around at the time PID 1 processes the message, or if the sending process is explicitly runtime-tracked by the service manager. The latter is the case if the service manager originally forked off the process, i.e. on all processes that match NotifyAccess=main or NotifyAccess=exec. Conversely, if an auxiliary process of the unit sends an sd_notify() message and immediately exits, the service manager might not be able to properly attribute the message to the unit, and thus will ignore it, even if NotifyAccess=all is set for it.

dbus-activated units spawned by us satisfy the second condition

**Review by yuwata — CHANGES_REQUESTED** [review]
I am not sure if this is the way to go, but there are many places "STOPPING=1" is sent. Should we also call sd_notify_barrier() after all "STOPPING=1" notifications?

**Comment by Geass-LL** [comment]
> I am not sure if this is the way to go, but there are many places "STOPPING=1" is sent. Should we also call sd_notify_barrier() after all "STOPPING=1" notifications?

No. The issue occurs when the service exits, although other places use "STOPPING=1", they are resident processes, and don't exit frequently. systemd-timedated, systemd-hostnamed, systemd-localed, ... which use "bus_event_loop_with_idle" are not, they are activated by dbus, and will exit when they are idle.

**Comment by Geass-LL** [comment]
> Why is this needed? From `sd_notify(3)`:
> 
> > Note that sd_notify() notifications may be attributed to units correctly only if either the sending process is still around at the time PID 1 processes the message, or if the sending process is explicitly runtime-tracked by the service manager. The latter is the case if the service manager originally forked off the process, i.e. on all processes that match NotifyAccess=main or NotifyAccess=exec. Conversely, if an auxiliary process of the unit sends an sd_notify() message and immediately exits, the service manager might not be able to properly attribute the message to the unit, and thus will ignore it, even if NotifyAccess=all is set for it.
> 
> dbus-activated units spawned by us satisfy the second condition

Does "the second condition" you methioned  means "the sending process is explicitly runtime-tracked by the service manager. "? When systemd gets the notify message, it checks which service the message from by 1. checking the watch_pid, 2. checking the process's cgroup. So if a different service forks a same PID process, systemd may calculate a  wrong service by 2.

**Comment by YHNdnzj** [comment]
Note that in v255, pid1 has been switched to pidfd for most process tracking. So this should no longer be a problem.

**Comment by Geass-LL** [comment]
> Note that in v255, pid1 has been switched to pidfd for most process tracking. So if your analysis is correct, it should no longer be a problem.

I don't think so. `manager_dispatch_notify_fd` initialize pidref by an invalid fd.

```c
  /* Generate lookup key from the PID (we have no pidfd here, after all) */
  PidRef pidref = PIDREF_MAKE_FROM_PID(ucred->pid);
.......
#define PIDREF_MAKE_FROM_PID(x) (PidRef) { .pid = (x), .fd = -EBADF }
```

This fd is useless, the verification will be skipped in `pidref_verify`, so I think the latest version also suffers this problem.

**Pull request closed without merging**